configs/parser_config.go

```diff
@@ -32,7 +32,7 @@ func (p *Parser) LoadConfigFileOverride(path string) (*File, hcl.Diagnostics) {
 func (p *Parser) loadConfigFile(path string, override bool) (*File, hcl.Diagnostics) {
 
 	body, diags := p.LoadHCLFile(path)
-	if body == nil {
+	if body == nil || diags.HasErrors() {
```

Comment:
While this would prevent the panic reported, it would also mean that any syntax error would prevent any analysis of the partially-decoded configuration: we're intentionally continuing here so that we can eventually return a partial `*File` object that can then be used for analysis use-cases, like in [terraform-lsp](https://github.com/juliosueiras/terraform-lsp) where it often needs to do analysis on incomplete configuration while the user is typing.

It looks like HCL is dealing with this syntax error by setting the value of the attribute to `cty.DynamicVal`. Therefore I think we could avoid the panic while retaining the partial-analysis capability by adjusting `sniffCoreVersionRequirements` to be more careful in how it accesses the `required_version` attribute... since errors are already being returned anyway in that case, I imagine it could just skip any block that has an unknown value for `required_version`.

What do you think? I'd rather avoid breaking these analysis use-cases if possible, since I expect that in future Terraform CLI itself will start to incorporate functionality like we see in [terraform-lsp](https://github.com/juliosueiras/terraform-lsp) today.

Reply:
Thanks for the explanation around why the `configs` package should continue on error, that wasn't clear. 

What I don't follow is why we would complete this processing during init/plan/apply when the syntax is invalid. I would expect an immediate error in that case; might it make sense to have the CLI commands do an initial pass through `hclsyntax.ParseConfig` so we can give the user an opportunity to fix the immediate errors before trying again?

Reply:
The philosophy here was to try to gather as many errors as possible to return all at once. The HCL parser is built to recover when it encounters an error, so a syntax error in one part of a file may not prevent us from parsing some other part of the file and generating a semantic error about it.

For example, if you wrote something like this:

```hcl
resource "foo" "bar" {
  lifecycle {}
  lifecycle {}
}

resource "invalid" "syntax" {
  nothing = 
}
```

I believe today it would return both the error about there being two `lifecycle` blocks (a semantic error) _and_ the error about `nothing = ` being invalid syntax.

As a user I tend to find it frustrating when a system requires me to fix one error at a time, only to run again and find another error. So this design was in the hope of minimizing the number of "round trips" through error handling. That can be particularly helpful if the error shows up in your automation system rather than at your local terminal prompt.

With all of that said, it might be that it's just too difficult to write our config loading logic to be tolerant of syntax errors, and I expect it isn't _too_ bad to force the user to deal with the syntax errors first and only then see the semantic errors. The one exception I would make for that is `terraform validate`, since it's primarily intended for use-cases like text editor integrations where it's helpful to draw red error lines under as many different things as possible, so that the user can work through them in order.

With that said, I'm not really sure if it's practical to make an exception for `terraform validate` and the language server without us actually fixing the issues that are causing these panics in the first place, because otherwise _those_ tools would still tend to panic on syntax errors and those ones are the main ones that tend to _encounter_ partial configurations that have syntax errors. :confounded:

Reply:
Follow-up thought here: when I was working on HCL I used [go-fuzz](https://github.com/dvyukov/go-fuzz) to track down various crashes caused by syntax errors in the HCL layer, by essentially just throwing random garbage at the parser to see whether it would fail or not.

I don't think we ever fuzz-tested the Terraform-layer config code though. Maybe we could root out a bunch of similar crasher issues here by adapting [HCL's fuzzing utilities](https://github.com/hashicorp/hcl2/tree/master/hcl/hclsyntax/fuzz) to work with Terraform's `configs` package instead. Based on the results of fuzz testing we will hopefully develop a better sense of how many crasher bugs there are and whether localized fixes for them are feasible; if localized fixes _are_ feasible then we can mark each one with a unit test so that we can detect any future maintenance regressions without having to run the fuzzer again.